### PR TITLE
Add tensorflow-quantum executor example to documentation

### LINKED ISSUE #334: Add executor example with TensorFlow Quantum
Issue Description
-----------------

An example with TensorFlow Quantum could be added to the executors' part of the user guide in the documentation.

### PR DESCRIPTION
**Pull request by k-m-schultz**
Description
-----------
Fixes #334 

Checklist
-----------
Please make sure you have finished the following tasks before requesting a review of the pull request (PR). For more information, please refer to the [Contributor's Guide](../blob/master/CONTRIBUTING.md):

- [x] Contributions to mitiq should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/). You can enforce it easily with [`black`][black] style and with [`flake8`][flake8] conventions.
- [x] Please add tests to cover your changes, if applicable, and make sure that all new and existing tests pass locally.
- [x] If the behavior of the code has changed or new feature has been added, please also [update the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md).
- [x] Functions and classes have useful [Google-style][google] docstrings and [type hints][hints] in the signature of the objects.
- [x] (Bug fix) The associated issue is referenced above using [auto-close keywords][auto-close]. If the PR fixes an issue, use the keyword fix/fixes/fixed followed by the issue id. For example, if the PR fixes issue 1184, type "Fixes #1184" (without quotes).
- [x] The [changelog][changelog] is updated, including author and PR number (@username, gh-xxx).

[auto-close]: https://help.github.com/en/articles/closing-issues-using-keywords
[black]: https://black.readthedocs.io/en/stable/index.html
[changelog]: https://github.com/unitaryfund/mitiq/blob/master/CHANGELOG.md
[contributing]: https://github.com/unitaryfund/mitiq/blob/master/CONTRIBUTING.md
[docs]: https://github.com/unitaryfund/mitiq/blob/master/docs/source/
[flake8]: http://flake8.pycqa.org
[google]: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
[hints]: https://www.python.org/dev/peps/pep-0484/
[pep-484]: https://www.python.org/dev/peps/pep-0484/

### CODE REVIEW

**Comment by nathanshammah** [comment]
I looked a bit into this as I was curious if just tweaking `doctest_global_setup` in the `conf.py` file would suffice, but indeed it seems that the libraries would need to be installed by the requirements anyways. 

Another problem is that [TF-quantum](https://github.com/tensorflow/quantum/issues/290) (and TF) does not support python 3.8, which is supported by Mitiq. My advice would be to forget the testing of this executor example and just use the `code-block` directive, at least for the time being. 

This is an example in which a separate `mitiq-examples` folder may be of help maybe, as it would allow a notebook to be installed with its dependencies. How the unit testing of notebooks works is not clear though.

**Comment by nathanshammah** [comment]
Just a note: the python 3.8 compatibility issue may be bypassed by installing [`tfq-nightly`](https://pypi.org/project/tfq-nightly/), which supports python 3.8. 

However, I cannot check it now, as something is messing with the sphinx doctests on my machine, which throw some 27 issues on circuits not being drawn as expected, as it downgraded cirq to v.0.8. Even by pip-uninstalling cirq, pip installing the 0.9dev version we pinned does not work properly for me, it may have messed something in my conda environment.

File: docs/source/guide/guide-executors.rst [hunk 1]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+
+    import numpy as np
+    from cirq import Circuit
+    import sympy
+
+    import tensorflow as tf
+    import tensorflow_quantum as tfq
+
+
+    def stochastic_bit_flip_simulation(circ: Circuit, p, num_MC=100):
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:452]: might as well add typing to the rest of this signature

File: docs/source/guide/guide-executors.rst [hunk 2]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        Simulates a circuit with random bit flip (X(\pi)) errors
+        Args:
+            circ: The quantum program as a cirq object
+            p: probability of an X(\pi) gate on each qubit after each gate in circ
+            num_MC: number of random trajectories to average over
+        Returns:
+            The expectation value of the 0 state as a float
+        """
+    
+        #Create array of symbolic variables and reshape to natural circuit parameterization
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:463]:
```suggestion
        # Create array of symbolic variables and reshape to natural circuit parameterization
```

File: docs/source/guide/guide-executors.rst [hunk 3]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        h = sympy.symbols(''.join(['h_{0} '.format(i) for i in range(len(circ.moments)*len(circ.all_qubits()))]),positive=True)
+        h_array = np.asarray(h).reshape((len(circ.all_qubits()),len(circ.moments)))
+
+        #Symbolicly add X gates to the input circuit
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:467]:
```suggestion
        # Symbolicly add X gates to the input circuit
```

File: docs/source/guide/guide-executors.rst [hunk 4]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        noisy_circuit = Circuit()
+        for i,moment in enumerate(circ.moments):
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:469]:
```suggestion
        for i, moment in enumerate(circ.moments):
```

File: docs/source/guide/guide-executors.rst [hunk 5]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+            noisy_circuit.append(moment)
+            for j,q in enumerate(circ.all_qubits()):
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:471]:
```suggestion
            for j, q in enumerate(circ.all_qubits()):
```

File: docs/source/guide/guide-executors.rst [hunk 6]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+                noisy_circuit.append(cirq.rx(h_array[j,i]).on(q))
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:472]:
```suggestion
                noisy_circuit.append(cirq.rx(h_array[j, i]).on(q))
```

File: docs/source/guide/guide-executors.rst [hunk 7]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        
+        #rotations will be pi w/ prob p, 0 w/ prob 1-p
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:474]:
```suggestion
        # rotations will be pi w/ prob p, 0 w/ prob 1-p
```

File: docs/source/guide/guide-executors.rst [hunk 8]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        vals = [np.reshape((np.random.rand(len(circ.all_qubits()),len(circ.moments))<p)*np.pi,(1,len(circ.all_qubits())*len(circ.moments))) for _ in range(num_MC)]
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:475]: maybe break this up so it's readable?

File: docs/source/guide/guide-executors.rst [hunk 9]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+            circ: The quantum program as a cirq object
+            p: probability of an X(\pi) gate on each qubit after each gate in circ
+            num_MC: number of random trajectories to average over
+        Returns:
+            The expectation value of the 0 state as a float
+        """
+    
+        #Create array of symbolic variables and reshape to natural circuit parameterization
+        h = sympy.symbols(''.join(['h_{0} '.format(i) for i in range(len(circ.moments)*len(circ.all_qubits()))]),positive=True)
+        h_array = np.asarray(h).reshape((len(circ.all_qubits()),len(circ.moments)))
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:465]: ditto here re: readability -- some breaking up & spaces would likely help

File: docs/source/guide/guide-executors.rst [hunk 10]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        vals = [np.reshape((np.random.rand(len(circ.all_qubits()),len(circ.moments))<p)*np.pi,(1,len(circ.all_qubits())*len(circ.moments))) for _ in range(num_MC)]
+        
+
+        
+        # needs to be a rank 2 tensor
+        vals = np.squeeze(vals)
+        if num_MC == 1:
+            vals = [vals]
+        
+        #Instantiate tfq layer for computing state vector
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:484]:
```suggestion
        # Instantiate tfq layer for computing state vector
```

File: docs/source/guide/guide-executors.rst [hunk 11]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        state = tfq.layers.State()
+        
+        #Execute monte carlo sim with symbolic values specified by vals
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:487]:
```suggestion
        # Execute monte carlo sim with symbolic values specified by vals
```

File: docs/source/guide/guide-executors.rst [hunk 12]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        out = state(noisy_circuit,symbol_names=h,symbol_values=vals).to_tensor()
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:488]:
```suggestion
        out = state(noisy_circuit, symbol_names=h, symbol_values=vals).to_tensor()
```

File: docs/source/guide/guide-executors.rst [hunk 13]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        
+        #Fancy way of computing and summing individual density operators, follwed by averaging
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:490]:
```suggestion
        # Fancy way of computing and summing individual density operators, followed by averaging
```

File: docs/source/guide/guide-executors.rst [hunk 14]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        dm = tf.tensordot(tf.transpose(out),tf.math.conj(out),axes=[[1],[0]]).numpy()/num_MC
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:491]:
```suggestion
        dm = tf.tensordot(tf.transpose(out), tf.math.conj(out), axes=[[1], [0]]).numpy() / num_MC
```

File: docs/source/guide/guide-executors.rst [hunk 15]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        
+        #return measurement of 0 state
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:493]:
```suggestion
        # return measurement of 0 state
```

File: docs/source/guide/guide-executors.rst [hunk 16]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        return np.real(dm[0,0])
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:494]:
```suggestion
        return np.real(dm[0, 0])
```

File: docs/source/guide/guide-executors.rst [hunk 17]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+this writing, TensorFlow Quantum is limited to
+
+  1. ``circ`` ``Circuits`` that use  ``circ`` ``GridQubit`` instances
+  2. Unitary circuits only, so non-unitary errors need to use Monte Carlo
+
+Despite this latter limitation, there is a crossover point where Monte Carlo using Tensorflow evaluates faster than the exact density matrix simulation using ``circ``.
+
+Below is an example to use TensorFlow Quantum to simulate a bit-flip channel:
+
+.. code::
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:442]: change back to testcode

File: docs/source/guide/guide-executors.rst [hunk 18]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+        #Execute monte carlo sim with symbolic values specified by vals
+        out = state(noisy_circuit,symbol_names=h,symbol_values=vals).to_tensor()
+        
+        #Fancy way of computing and summing individual density operators, follwed by averaging
+        dm = tf.tensordot(tf.transpose(out),tf.math.conj(out),axes=[[1],[0]]).numpy()/num_MC
+        
+        #return measurement of 0 state
+        return np.real(dm[0,0])
+
+.. testcode::
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:496]: can uncomment this after you make the changes in my top-level review regarding requirements

File: docs/source/guide/guide-executors.rst [hunk 19]
```diff
@@ -423,3 +423,87 @@ Qiskit: Hardware
...
+    
+    #import cirq
+    #from mitiq.benchmarks import randomized_benchmarking
+    #circ = randomized_benchmarking.rb_circuits(1,[20],1)[0]
+
+    ##Need to make sure the qubits are cirq.GridQubit
+    #circ=circ.transform_qubits(lambda q: cirq.GridQubit.rect(1,1)[0])
+
+    #out = stochastic_bit_flip_simulation(circ, .001)
+    #assert .97 < out < 1
```
Comment (karalekas) [inline docs/source/guide/guide-executors.rst:509]: this fails sometimes. i've seen values as low as 0.93999755. so this range will have to be extended

**Review by karalekas — CHANGES_REQUESTED** [review]
In addition to the inline comments (most of which are spacing related, but some of which are important like the assertion issue) this is what I suggest in order to make it so this PR will build and test the tfq code in the docs.

1. Add a new file `docs/requirements.txt` with two lines: `tensorflow~=2.3.1` and `tfq-nightly~=0.4.0.dev20201001`.

2. Install the requirements in this file when building the docs in `.github/workflows/ci.yml`. Specifically, there is a job named `docs` with a `run` entry in it. The run entry should have a new line added to it: `pip install -r docs/requirements.txt`. For extra clarity:

```yaml
- name: Install Python dependencies
        run: |
          python -m pip install --upgrade pip
          make install requirements
```

should change to

```yaml
- name: Install Python dependencies
        run: |
          python -m pip install --upgrade pip
          make install requirements
          pip install -r docs/requirements.txt
```

After making those two changes, and addressing the review comments, the build should work (it works for me locally).

Finally, make sure to add an entry to the changelog! Thanks for contributing! :)

**Note to self**: Make issue to add docs requirements installation instructions after this is merged.

**Review by nathanshammah — CHANGES_REQUESTED** [review]
Thank you @k-m-schultz. 
I added some comments requesting some changes:
- on the `docs/requirements.txt` file: adding information in the documentation, [here](https://github.com/unitaryfund/mitiq/blob/master/docs/CONTRIBUTING_DOCS.md#requirements), and merging the requirements for docs present in the main `requirements.txt` file. 
- Note that TF Quantum has a [0.4.0 release](https://github.com/tensorflow/quantum/releases/tag/0.4.0), that maybe works with what is needed? Better to ping a released version than a development one. I don't know if that will make conflicts arise with dependencies, as it requires numpy 1.5 and mitiq pins 1.5.
- a small suggestion: renaming `num_MC` to `num_montecarlo` or `montecarlo_repetitions` 
would be more explicit and slightly more "pythonic"
- It is nice the use of `sympy` symbols, this could be highlighted in the text, possibly.
- Most requests by @karalekas are implemented, but there are some outstanding comments to the code that need to be resolved or debated. Thanks!

File: docs/requirements.txt [hunk 20]
```diff
@@ -0,0 +1,2 @@
+tensorflow~=2.3.1
+tfq-nightly~=0.4.0.dev20201001
```
[the remark this answers is not in the document]
Comment (k-m-schultz) [reply docs/requirements.txt:2]: Locally, I verified this works with the tensorflow-quantum 0.4.0, so I will make that change to docs/requirements.  Not sure what to do about the other requirements from requirements.txt, let me know.

File: docs/source/guide/guide-executors.rst [hunk 21]
```diff
@@ -501,3 +501,83 @@ Qiskit: Hardware
...
+            circ: The quantum program as a cirq object
+            p: probability of an X(\pi) gate on each qubit after each gate in circ
+            num_monte_carlo: number of random trajectories to average over
+        Returns:
+            The expectation value of the 0 state as a float
+        """
+    
+        # Create array of symbolic variables and reshape to natural circuit parameterization
+        h = sympy.symbols(''.join(['h_{0} '.format(i) for i in range(len(circ.moments)*len(circ.all_qubits()))]),positive=True)
+        h_array = np.asarray(h).reshape((len(circ.all_qubits()),len(circ.moments)))
```
Comment (nathanshammah) [inline docs/source/guide/guide-executors.rst:543]:
```suggestion
        total_range = len(circ.moments) * len(circ.all_qubits())
        for i in range(total_range):
           to_join = 'h_{0} '.format(i) 
        h = sympy.symbols(''.join([to_join]), positive = True)
        h_array = np.asarray(h).reshape((len(circ.all_qubits()), len(circ.moments)))
```

**Review by nathanshammah — COMMENTED** [review]
LGTM, thanks @k-m-schultz! I made a small refactoring of the first two lines of the simulation to address Peter's comment.

**Review by karalekas — APPROVED** [review]

**Comment by karalekas** [comment]
Thanks for your contribution @k-m-schultz! :)

**Comment by nathanshammah** [comment]
Thanks @karalekas.

**Comment by crazy4pi314** [comment]
@all-contributors please add @k-m-schultz for example

**Pull request merged**